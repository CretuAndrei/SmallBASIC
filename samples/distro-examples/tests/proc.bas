
dim va(4)

for i=0 to 4
  va(i)=i
next

' array copy
vn()=va()
for i=0 to 4
	if ( va(i) <> vn(i) )
		throw "array copy error"
	fi
next

' ==== procedures

proc1
proc2 "String test"
proc2 1/2
proc3 1,2
proc4 4

i=-1
proc5 va(),4
if i<>-1 then ? "PROC5 failed" else ? "proc5: ok"

i=-1
proc6 i
if i<>6 then ? "PROC6 failed" else ? "proc6: ok"

proc6 va(1)
if va(1)<>6 then ? "PROC6 failed" else ? "proc6: ok"

' recursion
i=1
rec i

'==== functions

? "f(x)=";f(8)
x=8
? f(1)+1+sqr(f(x))+4

'==== return array!
vn()=fillarray
for i=0 to ubound(vn)
  ? vn(i)
next

' passing arrays byref
fillarr2 vn()
for i=0 to ubound(vn)
  ? vn(i)
next

' ===
dim vv(4)
vv()=fill3(vv())

' new code
vv=fill4(vv)

end

' =====================================

' Simple procedure
sub proc1
? "Proc1: no parameters"
end

' 1 parameter procedure
sub proc2(a)
? "Proc2=";a
end

' 2 parameter procedure
sub proc3(a,b)
? "Proc3=";a+b
end

' local procedures
sub proc4(a)

  sub proc5(b)

    sub proc6(b)
      ? "sub-proc6=";b
    end

  ? "sub-proc5=";b
  proc6 b
  end

? "Proc4=";a
proc5 a:REM local sub proc5 has priority
end

' null-arrays as parameter
sub proc5(x(),count)
local i

for i=0 to count
  ? "Proc5 - Array="; x(i)
next
end

' byref
sub proc6(byref n)
n=6
end

func f(x)
f=2^x
end

sub rec(byref n)
local s

n=n+1
s=n
? "R:";n
if n<10
  rec n
fi
? "REXIT:";s
end

func fillarray
local v,i

dim v(16)

for i=0 to 16
	v(i) = 16-i
next
fillarray=v()
end

sub fillarr2(byref v())
local i

for i=0 to 16
	v(i) = i
next
end

func fill3(v())
v(0)=1
fill3=v()
end

func fill4(v)
v(0)=1
fill4=v
end


