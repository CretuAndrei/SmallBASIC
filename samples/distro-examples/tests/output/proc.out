Proc1: no parameters
Proc2=String test
Proc2=0.5
Proc3=3
Proc4=4
sub-proc5=4
sub-proc6=4
Proc5 - Array=0
Proc5 - Array=1
Proc5 - Array=2
Proc5 - Array=3
Proc5 - Array=4
proc5: ok
proc6: ok
proc6: ok
R:2
R:3
R:4
R:5
R:6
R:7
R:8
R:9
R:10
REXIT:10
REXIT:9
REXIT:8
REXIT:7
REXIT:6
REXIT:5
REXIT:4
REXIT:3
REXIT:2
f(x)=256
23
16
15
14
13
12
11
10
9
8
7
6
5
4
3
2
1
0
0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
